reference-implementation/lib/ReadableStream-impl.js

```diff
@@ -0,0 +1,124 @@
+const aos = require('./abstract-ops/readable-streams.js');
+const wsAOs = require('./abstract-ops/writable-streams.js');
+
+const idlUtils = require('../generated/utils.js');
+const UnderlyingSource = require('../generated/UnderlyingSource.js');
+
+exports.implementation = class ReadableStreamImpl {
+  constructor(globalObject, [underlyingSource, strategy]) {
+    if (underlyingSource === undefined) {
+      underlyingSource = null;
+    }
+    const underlyingSourceDict = UnderlyingSource.convert(underlyingSource);
+
+    aos.InitializeReadableStream(this);
+
+    if (underlyingSourceDict.type === 'bytes') {
+      if ('size' in strategy) {
+        throw new RangeError('The strategy for a byte stream cannot have a size function');
+      }
+
+      const highWaterMark = ExtractHighWaterMark(strategy, 0);
+      aos.SetUpReadableByteStreamControllerFromUnderlyingSource(
+        this, underlyingSource, underlyingSourceDict, highWaterMark
+      );
+    } else {
+      assert(!('type' in underlyingSourceDict));
+      const sizeAlgorithm = ExtractSizeAlgorithm(strategy);
+      const highWaterMark = ExtractHighWaterMark(strategy, 1);
+      aos.SetUpReadableStreamDefaultControllerFromUnderlyingSource(
+        this, underlyingSource, underlyingSourceDict, highWaterMark, sizeAlgorithm
+      );
+    }
+  }
+
+  get locked() {
+    return aos.IsReadableStreamLocked(this);
+  }
+
+  cancel(reason) {
+    if (aos.IsReadableStreamLocked(this) === true) {
+      return promiseRejectedWith(new TypeError('Cannot cancel a stream that already has a reader'));
+    }
+
+    return aos.ReadableStreamCancel(this, reason);
+  }
+
+  getReader(options) {
+    if (!('mode' in options)) {
+      return aos.AcquireReadableStreamDefaultReader(this, true);
+    }
+
+    assert(options.mode === 'byob');
+    return aos.AcquireReadableStreamBYOBReader(this, true);
+  }
+
+  pipeThrough(transform, options) {
+    // Conversion here is needed until https://github.com/jsdom/webidl2js/issues/81 is fixed.
+    if ('signal' in options) {
+      if (!isAbortSignal(options.signal)) {
+        return promiseRejectedWith(new TypeError('Invalid signal argument'));
+      }
+    }
+
+    if (aos.IsReadableStreamLocked(this) === true) {
+      throw new TypeError('ReadableStream.prototype.pipeThrough cannot be used on a locked ReadableStream');
+    }
+    if (wsAOs.IsWritableStreamLocked(transform.writable) === true) {
+      throw new TypeError('ReadableStream.prototype.pipeThrough cannot be used on a locked WritableStream');
+    }
+
+    const promise = aos.ReadableStreamPipeTo(
+      this, transform.writable, options.preventClose, options.preventAbort, options.preventCancel, options.signal
+    );
+
+    setPromiseIsHandledToTrue(promise);
+
+    return transform.readable;
+  }
+
+  pipeTo(destination, options) {
+    // Conversion here is needed until https://github.com/jsdom/webidl2js/issues/81 is fixed.
```

Comment:
I find the word "conversion" confusing here. Would "type checking" be acceptable instead?